File: 8/day8.js
const fs = require('fs');

function antinodeCount(antenas, rows, cols, start, n) {
  antinodes = new Set();
  for (const pos of Object.values(antenas)) {
    for (let i = 0; i < pos.length; i++) {
      for (let j = i + 1; j < pos.length; j++) {
        let [ri, ci] = pos[i];
        let [rj, cj] = pos[j];
        let [dr, dc] = [ri - rj, ci - cj]
        for (let k = start; k <= n; k++) {
          [ari, aci] = [ri + k*dr, ci + k*dc]
          if (0 <= ari && ari < rows && 0 <= aci && aci < cols) antinodes.add(`${ari},${aci}`);
          [arj, acj] = [rj - k*dr, cj - k*dc]
          if (0 <= arj && arj < rows && 0 <= acj && acj < cols) antinodes.add(`${arj},${acj}`);
        }
      }
    }
  }
  console.log(antinodes.size);
}
function main(filePath) {
    const grid = fs.readFileSync(filePath, 'utf-8')
      .split('\n')
      .map(line => line.trim())
      .filter(line => line !== '');
    // console.log(grid)
    const rows = grid.length;
    const cols = grid[0].length;
    antenas = {};
    for (let r = 0; r < rows; r++) {
      for (let c = 0; c < cols; c++) {
        char = grid[r][c];
        if (char === '.') continue;
        if (!antenas[char]) antenas[char] = [];
        antenas[char].push([r, c]);
      }
    }
    console.log(antinodeCount(antenas, rows, cols, 1, 1));
    console.log(antinodeCount(antenas, rows, cols, 0, rows));   
}

// main('./8/input0.txt')
main('./8/input.txt')

File: 8/day8.py
from pprint import pp
def main(file_path: str) -> None:
    with open(file_path) as f:
        grid = f.read().splitlines()
    # pp(grid)
    rows, cols = len(grid), len(grid[0])
    
    antenas = {}
    for r in range(rows):
        for c in range(cols):
            char = grid[r][c]
            if char != '.':
                antenas[char] = antenas.get(char, []) + [(r, c)]
    
    def antinode_count(start: int, n: int = 1) -> None:
        antinodes = set()
        for pos in antenas.values():
            for i in range(len(pos)-1):
                for j in range(i+1, len(pos)):
                    ri, ci = pos[i]
                    rj, cj = pos[j]
                    dr, dc = ri - rj, ci - cj
                    for k in range(start, n+1):
                        if 0 <= ri + k*dr < rows and 0 <= ci + k*dc < cols:
                            antinodes.add((ri + k*dr, ci + k*dc))
                        if 0 <= rj - k*dr < rows and 0 <= cj - k*dc < cols:
                            antinodes.add((rj - k*dr, cj - k*dc))
        print(len(antinodes))
    antinode_count(start=1, n=1)
    # starts from 0 because we also count the 2 antenas
    antinode_count(start=0, n=rows)
    
    
if __name__ == '__main__':
    # main('./8/input0.txt')
    main('./8/input.txt')
